src/source/amqp.rs

```diff
@@ -0,0 +1,236 @@
+
+use crate::source::prelude::*;
+use serde::{Deserialize};
+use crate::url::TremorUrl;
+use url::Url;
+use crate::errors::Error;
+use lapin::{
+    options::*, types::FieldTable, Connection, ConnectionProperties, Consumer,
+};
+
+/*enum QueueProperties {
+    Durable 1
+    Exclusive 2
+    AutoDelete 4
+    NoWait 8
+}*/
+#[derive(Deserialize, Debug, Clone)]
+#[allow(clippy::struct_excessive_bools)]
+pub struct Config {
+    pub amqp_addr: String,
+    queue_name: String,
+    queue_options: QueueDeclareOptions,
+    routing_key: String,
+    exchange: String,  // "", by default
+    #[serde(default = "Default::default")]
+    pub close_on_done: bool,
+    #[serde(default = "Default::default")]
+    pub sleep_on_done: u64,
+}
+
+impl ConfigImpl for Config {}
+
+pub struct Amqp {
+    pub config: Config,
+    onramp_id: TremorUrl,
+}
+
+pub struct Int {
+    uid: u64,
+    config: Config,
+    amqp_url: Url,
+    onramp_id: TremorUrl,
+    origin_uri: EventOriginUri,
+    consumer: Option<Consumer>,
+    with_ack: bool,
+}
+
+impl std::fmt::Debug for Int {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Kafka")
+    }
+}
+
+impl onramp::Impl for Amqp {
+    fn from_config(id: &TremorUrl, config: &Option<YamlValue>) -> Result<Box<dyn Onramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            Ok(Box::new(Self {
+                config,
+                onramp_id: id.clone(),
+            }))
+        } else {
+            Err("Missing config for Amqp onramp".into())
+        }
+    }
+}
+
+impl Int {
+    fn from_config(uid: u64, onramp_id: TremorUrl, config: &Config) -> Self {
+        let amqp_url = Url::parse(&config.amqp_addr).unwrap();
+        let origin_uri = EventOriginUri {
+            uid,
+            scheme: "amqp".to_string(),
+            host: "not-connected".to_string(),
+            port: None,
+            path: vec![],
+        };
+        Self {
+            uid,
+            config: config.clone(),
+            amqp_url,
+            onramp_id,
+            consumer: None,
+            origin_uri,
+            with_ack: false,
+        }
+    }
+}
+
+impl std::fmt::Debug for Amqp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Amqp")
+    }
+}
+
+impl std::convert::From<lapin::Error> for Error {
+    fn from(e: lapin::Error) -> Self {
+        Self::from(format!("{:?}", e))
+    }
+}
+
+#[async_trait::async_trait()]
+impl Source for Int {
+    fn is_transactional(&self) -> bool {
+        self.with_ack
+    }
+    fn id(&self) -> &TremorUrl {
+        &self.onramp_id
+    }
+    async fn pull_event(&mut self, _id: u64) -> Result<SourceReply> {
+        if self.consumer.is_none() {
+            return Ok(SourceReply::StateChange(SourceState::Disconnected))
+        }
+        if let Some(delivery) = self.consumer.as_mut().unwrap().next().await {
```

Comment:
we don't allow unwrap or expect, instead we would prefer errors bubbling up in this cases.

The reasoning here is that if a single source or sink has an issue we'd not want to panic the entire application and potentially take down unrelated and pipelines that themselves are unaffected. This gets more important since we allow loading configuration at run time, and someone making an API call could, if we'd allow unwrap/expect/panic/etc. then crash the application with providing a 'bad' configuration

Reply:
Resolved it into a match tree